# Pandas Assignment - Arif Aygun

In [1]:
import pandas as pd

### 1- How many rows of data are in the DataFrame?

Please refer to the following link for this practice:

https://tf-assets-prod.s3.amazonaws.com/tf-curric/data-science/ksprojects.csv

In [2]:
ks_project = pd.read_csv(r"C:\Users\AAygun\Documents\GitHub\Data Sets\Magnimind\ksprojects.csv")
ks_project.head()

id                                               name  \
0  1147015301  "All We Had" Gets Into Cannes -- $10 or More G...   
1  1100844465                                       daily digest   
2  1316334968                             Drive A Faster Car 2.0   
3  1304906577         Accidental to Edinburgh - PHASE 1: AIRFARE   
4  1099226462          Logical Guess Pictures' 2nd Horror Movie!   

       category main_category             deadline    goal  \
0   Documentary  Film & Video  2009-05-20 21:50:00   300.0   
1   Documentary  Film & Video  2009-06-01 02:20:00   700.0   
2    Nonfiction    Publishing  2009-06-04 06:00:00  1000.0   
3       Theater       Theater  2009-06-05 05:59:00  6000.0   
4  Film & Video  Film & Video  2009-06-06 00:45:00   500.0   

              launched       state  backers country  usd_pledged currency  \
0  2009-04-30 22:10:30      failed        4      US        40.00      USD   
1  2009-05-04 21:14:28  successful       14      US       700.00      USD   
2  2009-05-04 22:51:31  successful       32      US      1367.00      USD   
3  2009-04-30 22:22:43  successful       24      US      6575.00      USD   
4  2009-04-30 01:32:55  successful       22      US       501.66      USD   

   pledged  
0    40.00  
1   700.00  
2  1367.00  
3  6575.00  
4   501.66

In [3]:
ks_project.shape

(65418, 13)

DataFrame has 65418 rows

### 2- What are the names and data types of the columns?

In [4]:
print(ks_project.columns)

Index(['id', 'name', 'category', 'main_category', 'deadline', 'goal',
       'launched', 'state', 'backers', 'country', 'usd_pledged', 'currency',
       'pledged'],
      dtype='object')


In [5]:
print(ks_project.dtypes)

id                 int64
name              object
category          object
main_category     object
deadline          object
goal             float64
launched          object
state             object
backers            int64
country           object
usd_pledged      float64
currency          object
pledged          float64
dtype: object


In [6]:
print(type(ks_project))

<class 'pandas.core.frame.DataFrame'>


### 3- Do any of the columns contain null values?

In [7]:
ks_project.isnull().values.any()

True

In [8]:
ks_project.isnull()

id   name  category  main_category  deadline   goal  launched  \
0      False  False     False          False     False  False     False   
1      False  False     False          False     False  False     False   
2      False  False     False          False     False  False     False   
3      False  False     False          False     False  False     False   
4      False  False     False          False     False  False     False   
...      ...    ...       ...            ...       ...    ...       ...   
65413  False  False     False          False     False  False     False   
65414  False  False     False          False     False  False     False   
65415  False  False     False          False     False  False     False   
65416  False  False     False          False     False  False     False   
65417  False  False     False          False     False  False     False   

       state  backers  country  usd_pledged  currency  pledged  
0      False    False    False        False     False    False  
1      False    False    False        False     False    False  
2      False    False    False        False     False    False  
3      False    False    False        False     False    False  
4      False    False    False        False     False    False  
...      ...      ...      ...          ...       ...      ...  
65413  False    False    False        False     False    False  
65414  False    False    False        False     False    False  
65415  False    False    False        False     False    False  
65416  False    False    False        False     False    False  
65417  False    False    False        False     False    False  

[65418 rows x 13 columns]

### 4- Find all successful documentary projects and sort them by the amount pledged. Print the top 10 highest pledges.

In [9]:
ks_project['main_category']

0        Film & Video
1        Film & Video
2          Publishing
3             Theater
4        Film & Video
             ...     
65413      Technology
65414      Technology
65415          Design
65416      Technology
65417          Design
Name: main_category, Length: 65418, dtype: object

In [10]:
ks_project['category']

0           Documentary
1           Documentary
2            Nonfiction
3               Theater
4          Film & Video
              ...      
65413             Sound
65414           Gadgets
65415    Product Design
65416        Technology
65417    Product Design
Name: category, Length: 65418, dtype: object

In [11]:
success_doc_pro = ks_project[(ks_project["category"] == "Documentary") & (ks_project["state"] == "successful")]
print(success_doc_pro)

               id                                               name  \
1      1100844465                                       daily digest   
9      1125367305            Meeting Andrei Tarkovsky - Film Website   
24     1216330417  Do It Again: One Man's Quest to Reunite the Kinks   
39     1039240602       The Film Dailies - Live From The Walnut Room   
40     1211984590                               'Mister Rogers & Me'   
...           ...                                                ...   
64092  1205743261         AgriCultured: The Lives of Organic Farmers   
64096  1334836761                           The Last Ring Home on TV   
64117  1175698090                                 BlackBook Medellin   
64129  1040239768  The Protestors at Standing Rock - Their Testim...   
64259  1287371551                                   Connections Film   

          category main_category             deadline     goal  \
1      Documentary  Film & Video  2009-06-01 02:20:00    700.0   
9  

In [12]:
success_doc_pro

id                                               name  \
1      1100844465                                       daily digest   
9      1125367305            Meeting Andrei Tarkovsky - Film Website   
24     1216330417  Do It Again: One Man's Quest to Reunite the Kinks   
39     1039240602       The Film Dailies - Live From The Walnut Room   
40     1211984590                               'Mister Rogers & Me'   
...           ...                                                ...   
64092  1205743261         AgriCultured: The Lives of Organic Farmers   
64096  1334836761                           The Last Ring Home on TV   
64117  1175698090                                 BlackBook Medellin   
64129  1040239768  The Protestors at Standing Rock - Their Testim...   
64259  1287371551                                   Connections Film   

          category main_category             deadline     goal  \
1      Documentary  Film & Video  2009-06-01 02:20:00    700.0   
9      Documentary  Film & Video  2009-06-21 02:31:00   1250.0   
24     Documentary  Film & Video  2009-08-10 21:26:00   5500.0   
39     Documentary  Film & Video  2009-09-15 01:00:00    500.0   
40     Documentary  Film & Video  2009-09-20 05:59:00  10000.0   
...            ...           ...                  ...      ...   
64092  Documentary  Film & Video  2016-11-27 22:36:26   1000.0   
64096  Documentary  Film & Video  2016-11-28 00:59:00   6000.0   
64117  Documentary  Film & Video  2016-11-28 19:27:52   5000.0   
64129  Documentary  Film & Video  2016-11-29 08:00:00   3000.0   
64259  Documentary  Film & Video  2016-12-01 23:28:43  25000.0   

                  launched       state  backers country  usd_pledged currency  \
1      2009-05-04 21:14:28  successful       14      US       700.00      USD   
9      2009-05-30 06:27:42  successful       60      US      2721.00      USD   
24     2009-05-17 05:55:13  successful       79      US      6632.32      USD   
39     2009-08-15 19:52:42  successful        8      US       730.00      USD   
40     2009-08-19 06:16:34  successful      176      US     10795.70      USD   
...                    ...         ...      ...     ...          ...      ...   
64092  2016-10-28 22:36:26  successful       10      US       500.00      USD   
64096  2016-10-20 07:56:50  successful      176      US      4055.00      USD   
64117  2016-10-19 19:27:52  successful       56      US      1106.00      USD   
64129  2016-11-13 20:15:35  successful       74      US         0.00      USD   
64259  2016-10-27 23:28:43  successful      179      US       300.00      USD   

        pledged  
1        700.00  
9       2721.00  
24      6632.32  
39       730.00  
40     10795.70  
...         ...  
64092   1000.00  
64096  40830.00  
64117   5437.00  
64129   5000.00  
64259  25581.00  

[1087 rows x 13 columns]

In [13]:
success_doc_pro.sort_values(by=['usd_pledged'])

id                                               name  \
64129  1040239768  The Protestors at Standing Rock - Their Testim...   
63047  1077980678                              Lives of The Homeless   
62397  1073535790                                    London Recruits   
31857  1025733989                                           Dominoes   
61581   100358406                                 Charlie vs Goliath   
...           ...                                                ...   
28003  1205913248                                     ILEANA'S SMILE   
12127  1326363492                                   The Culture High   
22237   122290911                          The Killing of Tony Blair   
24166  1162836718  "The Wrecking Crew" The Untold Story of Rock &...   
11731  1078982332                BRIDEGROOM - An American Love Story   

          category main_category             deadline      goal  \
64129  Documentary  Film & Video  2016-11-29 08:00:00    3000.0   
63047  Documentary  Film & Video  2016-10-25 00:51:09     100.0   
62397  Documentary  Film & Video  2016-10-02 01:00:00    8000.0   
31857  Documentary  Film & Video  2014-08-21 03:25:28      10.0   
61581  Documentary  Film & Video  2016-09-02 02:00:00   40000.0   
...            ...           ...                  ...       ...   
28003  Documentary  Film & Video  2014-05-21 09:05:00  133000.0   
12127  Documentary  Film & Video  2012-08-04 04:00:00  190000.0   
22237  Documentary  Film & Video  2013-10-07 18:17:14   50000.0   
24166  Documentary  Film & Video  2013-12-21 08:59:00  250000.0   
11731  Documentary  Film & Video  2012-07-20 01:00:28  300000.0   

                  launched       state  backers country  usd_pledged currency  \
64129  2016-11-13 20:15:35  successful       74      US        0.000      USD   
63047  2016-10-04 00:51:09  successful        1      US        0.000      USD   
62397  2016-09-08 09:02:17  successful      103      GB        0.000      GBP   
31857  2014-08-06 03:25:28  successful        6      CA       18.312      CAD   
61581  2016-08-03 21:54:32  successful      267      US       25.000      USD   
...                    ...         ...      ...     ...          ...      ...   
28003  2014-04-21 09:05:00  successful     1173      US   177421.000      USD   
12127  2012-06-22 01:40:20  successful     3448      US   240022.000      USD   
22237  2013-08-28 18:17:14  successful     4459      GB   255150.000      GBP   
24166  2013-10-30 21:06:34  successful     4245      US   313158.000      USD   
11731  2012-06-20 01:00:28  successful     6508      US   384376.000      USD   

        pledged  
64129    5000.0  
63047     100.0  
62397    8603.0  
31857      20.0  
61581   40546.0  
...         ...  
28003  177421.0  
12127  240022.0  
22237  163891.0  
24166  313158.0  
11731  384376.0  

[1087 rows x 13 columns]

In [14]:
pledged_amount = success_doc_pro.sort_values('usd_pledged', ascending=False)
pledged_amount

id                                               name  \
11731  1078982332                BRIDEGROOM - An American Love Story   
24166  1162836718  "The Wrecking Crew" The Untold Story of Rock &...   
22237   122290911                          The Killing of Tony Blair   
12127  1326363492                                   The Culture High   
28003  1205913248                                     ILEANA'S SMILE   
...           ...                                                ...   
61581   100358406                                 Charlie vs Goliath   
31857  1025733989                                           Dominoes   
63047  1077980678                              Lives of The Homeless   
62397  1073535790                                    London Recruits   
64129  1040239768  The Protestors at Standing Rock - Their Testim...   

          category main_category             deadline      goal  \
11731  Documentary  Film & Video  2012-07-20 01:00:28  300000.0   
24166  Documentary  Film & Video  2013-12-21 08:59:00  250000.0   
22237  Documentary  Film & Video  2013-10-07 18:17:14   50000.0   
12127  Documentary  Film & Video  2012-08-04 04:00:00  190000.0   
28003  Documentary  Film & Video  2014-05-21 09:05:00  133000.0   
...            ...           ...                  ...       ...   
61581  Documentary  Film & Video  2016-09-02 02:00:00   40000.0   
31857  Documentary  Film & Video  2014-08-21 03:25:28      10.0   
63047  Documentary  Film & Video  2016-10-25 00:51:09     100.0   
62397  Documentary  Film & Video  2016-10-02 01:00:00    8000.0   
64129  Documentary  Film & Video  2016-11-29 08:00:00    3000.0   

                  launched       state  backers country  usd_pledged currency  \
11731  2012-06-20 01:00:28  successful     6508      US   384376.000      USD   
24166  2013-10-30 21:06:34  successful     4245      US   313158.000      USD   
22237  2013-08-28 18:17:14  successful     4459      GB   255150.000      GBP   
12127  2012-06-22 01:40:20  successful     3448      US   240022.000      USD   
28003  2014-04-21 09:05:00  successful     1173      US   177421.000      USD   
...                    ...         ...      ...     ...          ...      ...   
61581  2016-08-03 21:54:32  successful      267      US       25.000      USD   
31857  2014-08-06 03:25:28  successful        6      CA       18.312      CAD   
63047  2016-10-04 00:51:09  successful        1      US        0.000      USD   
62397  2016-09-08 09:02:17  successful      103      GB        0.000      GBP   
64129  2016-11-13 20:15:35  successful       74      US        0.000      USD   

        pledged  
11731  384376.0  
24166  313158.0  
22237  163891.0  
12127  240022.0  
28003  177421.0  
...         ...  
61581   40546.0  
31857      20.0  
63047     100.0  
62397    8603.0  
64129    5000.0  

[1087 rows x 13 columns]

In [15]:
pledged_amount.head(10)

id                                               name  \
11731  1078982332                BRIDEGROOM - An American Love Story   
24166  1162836718  "The Wrecking Crew" The Untold Story of Rock &...   
22237   122290911                          The Killing of Tony Blair   
12127  1326363492                                   The Culture High   
28003  1205913248                                     ILEANA'S SMILE   
27402  1135999291                             Rewind To Fast-Forward   
16243  1061853312                               Linsanity: the Movie   
27268  1228065368                  El Voto Hispano | The Documentary   
43820  1194449879  PlantPure Nation, a Documentary with Dr. T. Co...   
55322  1264072338                                   BREXIT THE MOVIE   

          category main_category             deadline      goal  \
11731  Documentary  Film & Video  2012-07-20 01:00:28  300000.0   
24166  Documentary  Film & Video  2013-12-21 08:59:00  250000.0   
22237  Documentary  Film & Video  2013-10-07 18:17:14   50000.0   
12127  Documentary  Film & Video  2012-08-04 04:00:00  190000.0   
28003  Documentary  Film & Video  2014-05-21 09:05:00  133000.0   
27402  Documentary  Film & Video  2014-05-01 07:59:00  137000.0   
16243  Documentary  Film & Video  2013-02-16 00:00:00  117000.0   
27268  Documentary  Film & Video  2014-04-27 08:15:23  158000.0   
43820  Documentary  Film & Video  2015-05-14 17:41:41  150000.0   
55322  Documentary  Film & Video  2016-03-02 01:02:47  100000.0   

                  launched       state  backers country  usd_pledged currency  \
11731  2012-06-20 01:00:28  successful     6508      US     384376.0      USD   
24166  2013-10-30 21:06:34  successful     4245      US     313158.0      USD   
22237  2013-08-28 18:17:14  successful     4459      GB     255150.0      GBP   
12127  2012-06-22 01:40:20  successful     3448      US     240022.0      USD   
28003  2014-04-21 09:05:00  successful     1173      US     177421.0      USD   
27402  2014-03-30 13:35:31  successful     4395      US     176746.0      USD   
16243  2013-01-15 19:46:34  successful     2102      US     167917.0      USD   
27268  2014-03-18 07:15:23  successful      461      US     165401.0      USD   
43820  2015-03-15 16:41:41  successful     1099      US     163930.0      USD   
55322  2016-02-01 01:02:47  successful     1685      GB     162393.0      GBP   

        pledged  
11731  384376.0  
24166  313158.0  
22237  163891.0  
12127  240022.0  
28003  177421.0  
27402  176746.0  
16243  167917.0  
27268  165401.0  
43820  163930.0  
55322  114000.0

##### - The top 10 highest pledges.

### 5- Create a new column named average_per_backer and set its value to the total amount pledged / number of backers.

In [16]:
pledged_amount["average_per_backer"] = pledged_amount["usd_pledged"] / pledged_amount["backers"]
pledged_amount

id                                               name  \
11731  1078982332                BRIDEGROOM - An American Love Story   
24166  1162836718  "The Wrecking Crew" The Untold Story of Rock &...   
22237   122290911                          The Killing of Tony Blair   
12127  1326363492                                   The Culture High   
28003  1205913248                                     ILEANA'S SMILE   
...           ...                                                ...   
61581   100358406                                 Charlie vs Goliath   
31857  1025733989                                           Dominoes   
63047  1077980678                              Lives of The Homeless   
62397  1073535790                                    London Recruits   
64129  1040239768  The Protestors at Standing Rock - Their Testim...   

          category main_category             deadline      goal  \
11731  Documentary  Film & Video  2012-07-20 01:00:28  300000.0   
24166  Documentary  Film & Video  2013-12-21 08:59:00  250000.0   
22237  Documentary  Film & Video  2013-10-07 18:17:14   50000.0   
12127  Documentary  Film & Video  2012-08-04 04:00:00  190000.0   
28003  Documentary  Film & Video  2014-05-21 09:05:00  133000.0   
...            ...           ...                  ...       ...   
61581  Documentary  Film & Video  2016-09-02 02:00:00   40000.0   
31857  Documentary  Film & Video  2014-08-21 03:25:28      10.0   
63047  Documentary  Film & Video  2016-10-25 00:51:09     100.0   
62397  Documentary  Film & Video  2016-10-02 01:00:00    8000.0   
64129  Documentary  Film & Video  2016-11-29 08:00:00    3000.0   

                  launched       state  backers country  usd_pledged currency  \
11731  2012-06-20 01:00:28  successful     6508      US   384376.000      USD   
24166  2013-10-30 21:06:34  successful     4245      US   313158.000      USD   
22237  2013-08-28 18:17:14  successful     4459      GB   255150.000      GBP   
12127  2012-06-22 01:40:20  successful     3448      US   240022.000      USD   
28003  2014-04-21 09:05:00  successful     1173      US   177421.000      USD   
...                    ...         ...      ...     ...          ...      ...   
61581  2016-08-03 21:54:32  successful      267      US       25.000      USD   
31857  2014-08-06 03:25:28  successful        6      CA       18.312      CAD   
63047  2016-10-04 00:51:09  successful        1      US        0.000      USD   
62397  2016-09-08 09:02:17  successful      103      GB        0.000      GBP   
64129  2016-11-13 20:15:35  successful       74      US        0.000      USD   

        pledged  average_per_backer  
11731  384376.0           59.062077  
24166  313158.0           73.771025  
22237  163891.0           57.221350  
12127  240022.0           69.611949  
28003  177421.0          151.254049  
...         ...                 ...  
61581   40546.0            0.093633  
31857      20.0            3.052000  
63047     100.0            0.000000  
62397    8603.0            0.000000  
64129    5000.0            0.000000  

[1087 rows x 14 columns]

In [17]:
pledged_amount["usd_pledged"]

11731    384376.000
24166    313158.000
22237    255150.000
12127    240022.000
28003    177421.000
            ...    
61581        25.000
31857        18.312
63047         0.000
62397         0.000
64129         0.000
Name: usd_pledged, Length: 1087, dtype: float64

### 6- What happened to the rows with 0 backers? How can this be dealt with?

In [18]:
pledged_amount.shape

(1087, 14)

### 7- Drop all rows with 0 backers then repeat the previous exercise.

In [19]:
pledged_amount.dropna()

id                                               name  \
11731  1078982332                BRIDEGROOM - An American Love Story   
24166  1162836718  "The Wrecking Crew" The Untold Story of Rock &...   
22237   122290911                          The Killing of Tony Blair   
12127  1326363492                                   The Culture High   
28003  1205913248                                     ILEANA'S SMILE   
...           ...                                                ...   
61581   100358406                                 Charlie vs Goliath   
31857  1025733989                                           Dominoes   
63047  1077980678                              Lives of The Homeless   
62397  1073535790                                    London Recruits   
64129  1040239768  The Protestors at Standing Rock - Their Testim...   

          category main_category             deadline      goal  \
11731  Documentary  Film & Video  2012-07-20 01:00:28  300000.0   
24166  Documentary  Film & Video  2013-12-21 08:59:00  250000.0   
22237  Documentary  Film & Video  2013-10-07 18:17:14   50000.0   
12127  Documentary  Film & Video  2012-08-04 04:00:00  190000.0   
28003  Documentary  Film & Video  2014-05-21 09:05:00  133000.0   
...            ...           ...                  ...       ...   
61581  Documentary  Film & Video  2016-09-02 02:00:00   40000.0   
31857  Documentary  Film & Video  2014-08-21 03:25:28      10.0   
63047  Documentary  Film & Video  2016-10-25 00:51:09     100.0   
62397  Documentary  Film & Video  2016-10-02 01:00:00    8000.0   
64129  Documentary  Film & Video  2016-11-29 08:00:00    3000.0   

                  launched       state  backers country  usd_pledged currency  \
11731  2012-06-20 01:00:28  successful     6508      US   384376.000      USD   
24166  2013-10-30 21:06:34  successful     4245      US   313158.000      USD   
22237  2013-08-28 18:17:14  successful     4459      GB   255150.000      GBP   
12127  2012-06-22 01:40:20  successful     3448      US   240022.000      USD   
28003  2014-04-21 09:05:00  successful     1173      US   177421.000      USD   
...                    ...         ...      ...     ...          ...      ...   
61581  2016-08-03 21:54:32  successful      267      US       25.000      USD   
31857  2014-08-06 03:25:28  successful        6      CA       18.312      CAD   
63047  2016-10-04 00:51:09  successful        1      US        0.000      USD   
62397  2016-09-08 09:02:17  successful      103      GB        0.000      GBP   
64129  2016-11-13 20:15:35  successful       74      US        0.000      USD   

        pledged  average_per_backer  
11731  384376.0           59.062077  
24166  313158.0           73.771025  
22237  163891.0           57.221350  
12127  240022.0           69.611949  
28003  177421.0          151.254049  
...         ...                 ...  
61581   40546.0            0.093633  
31857      20.0            3.052000  
63047     100.0            0.000000  
62397    8603.0            0.000000  
64129    5000.0            0.000000  

[1087 rows x 14 columns]

In [20]:
pledged_amount.shape

(1087, 14)

### 8- Create a DataFrame whose index is the first 10 letters of the alphabet and contains two more columns with first 10 prime numbers and the first 10 fibonacci numbers.

In [21]:
index = list('abcdefghij')
primes = [2, 3, 5, 7, 11, 13, 17, 19, 23, 29]
fibonacci = [1, 1, 2, 3, 5, 8, 13, 21, 34, 55]

df = pd.DataFrame({"Primes": primes, "Fibonacci": fibonacci}, index = index)
print(df)

   Primes  Fibonacci
a       2          1
b       3          1
c       5          2
d       7          3
e      11          5
f      13          8
g      17         13
h      19         21
i      23         34
j      29         55


### 9- Merge these two DataFrames into a single DataFrame: nycflights13.

Please use the following dataset for merging purposes:

https://tf-assets-prod.s3.amazonaws.com/tf-curric/data-science/flights.csv

https://tf-assets-prod.s3.amazonaws.com/tf-curric/data-science/airlines.csv

In [22]:
flights = pd.read_csv(r"C:\Users\AAygun\Documents\GitHub\Data Sets\Magnimind\flights.csv")
airlines = pd.read_csv(r"C:\Users\AAygun\Documents\GitHub\Data Sets\Magnimind\airlines.csv")

In [23]:
flights

year  month  day  dep_time  sched_dep_time  dep_delay  arr_time  \
0       2013      1    1     517.0             515        2.0     830.0   
1       2013      1    1     533.0             529        4.0     850.0   
2       2013      1    1     542.0             540        2.0     923.0   
3       2013      1    1     544.0             545       -1.0    1004.0   
4       2013      1    1     554.0             600       -6.0     812.0   
...      ...    ...  ...       ...             ...        ...       ...   
336771  2013      9   30       NaN            1455        NaN       NaN   
336772  2013      9   30       NaN            2200        NaN       NaN   
336773  2013      9   30       NaN            1210        NaN       NaN   
336774  2013      9   30       NaN            1159        NaN       NaN   
336775  2013      9   30       NaN             840        NaN       NaN   

        sched_arr_time  arr_delay carrier  flight tailnum origin dest  \
0                  819       11.0      UA    1545  N14228    EWR  IAH   
1                  830       20.0      UA    1714  N24211    LGA  IAH   
2                  850       33.0      AA    1141  N619AA    JFK  MIA   
3                 1022      -18.0      B6     725  N804JB    JFK  BQN   
4                  837      -25.0      DL     461  N668DN    LGA  ATL   
...                ...        ...     ...     ...     ...    ...  ...   
336771            1634        NaN      9E    3393     NaN    JFK  DCA   
336772            2312        NaN      9E    3525     NaN    LGA  SYR   
336773            1330        NaN      MQ    3461  N535MQ    LGA  BNA   
336774            1344        NaN      MQ    3572  N511MQ    LGA  CLE   
336775            1020        NaN      MQ    3531  N839MQ    LGA  RDU   

        air_time  distance  hour  minute            time_hour  
0          227.0      1400     5      15  2013-01-01 05:00:00  
1          227.0      1416     5      29  2013-01-01 05:00:00  
2          160.0      1089     5      40  2013-01-01 05:00:00  
3          183.0      1576     5      45  2013-01-01 05:00:00  
4          116.0       762     6       0  2013-01-01 06:00:00  
...          ...       ...   ...     ...                  ...  
336771       NaN       213    14      55  2013-09-30 14:00:00  
336772       NaN       198    22       0  2013-09-30 22:00:00  
336773       NaN       764    12      10  2013-09-30 12:00:00  
336774       NaN       419    11      59  2013-09-30 11:00:00  
336775       NaN       431     8      40  2013-09-30 08:00:00  

[336776 rows x 19 columns]

In [24]:
airlines

carrier                         name
0       9E            Endeavor Air Inc.
1       AA       American Airlines Inc.
2       AS         Alaska Airlines Inc.
3       B6              JetBlue Airways
4       DL         Delta Air Lines Inc.
5       EV     ExpressJet Airlines Inc.
6       F9       Frontier Airlines Inc.
7       FL  AirTran Airways Corporation
8       HA       Hawaiian Airlines Inc.
9       MQ                    Envoy Air
10      OO        SkyWest Airlines Inc.
11      UA        United Air Lines Inc.
12      US              US Airways Inc.
13      VX               Virgin America
14      WN       Southwest Airlines Co.
15      YV           Mesa Airlines Inc.

In [26]:
nycflights13 = pd.merge(flights, airlines)
nycflights13

year  month  day  dep_time  sched_dep_time  dep_delay  arr_time  \
0       2013      1    1     517.0             515        2.0     830.0   
1       2013      1    1     533.0             529        4.0     850.0   
2       2013      1    1     554.0             558       -4.0     740.0   
3       2013      1    1     558.0             600       -2.0     924.0   
4       2013      1    1     558.0             600       -2.0     923.0   
...      ...    ...  ...       ...             ...        ...       ...   
336771  2013      9   19    1759.0            1805       -6.0    1946.0   
336772  2013      9   20    1758.0            1805       -7.0    1929.0   
336773  2013      9   22    1759.0            1805       -6.0    1945.0   
336774  2013      9   23    1759.0            1805       -6.0    1935.0   
336775  2013      9   24    1751.0            1805      -14.0    1937.0   

        sched_arr_time  arr_delay carrier  flight tailnum origin dest  \
0                  819       11.0      UA    1545  N14228    EWR  IAH   
1                  830       20.0      UA    1714  N24211    LGA  IAH   
2                  728       12.0      UA    1696  N39463    EWR  ORD   
3                  917        7.0      UA     194  N29129    JFK  LAX   
4                  937      -14.0      UA    1124  N53441    EWR  SFO   
...                ...        ...     ...     ...     ...    ...  ...   
336771            1953       -7.0      OO    5568  N760SK    LGA  CLE   
336772            1953      -24.0      OO    5568  N766SK    LGA  CLE   
336773            1953       -8.0      OO    5568  N772SK    LGA  CLE   
336774            1953      -18.0      OO    5568  N776SK    LGA  CLE   
336775            1953      -16.0      OO    5568  N785SK    LGA  CLE   

        air_time  distance  hour  minute            time_hour  \
0          227.0      1400     5      15  2013-01-01 05:00:00   
1          227.0      1416     5      29  2013-01-01 05:00:00   
2          150.0       719     5      58  2013-01-01 05:00:00   
3          345.0      2475     6       0  2013-01-01 06:00:00   
4          361.0      2565     6       0  2013-01-01 06:00:00   
...          ...       ...   ...     ...                  ...   
336771      64.0       419    18       5  2013-09-19 18:00:00   
336772      60.0       419    18       5  2013-09-20 18:00:00   
336773      67.0       419    18       5  2013-09-22 18:00:00   
336774      68.0       419    18       5  2013-09-23 18:00:00   
336775      67.0       419    18       5  2013-09-24 18:00:00   

                         name  
0       United Air Lines Inc.  
1       United Air Lines Inc.  
2       United Air Lines Inc.  
3       United Air Lines Inc.  
4       United Air Lines Inc.  
...                       ...  
336771  SkyWest Airlines Inc.  
336772  SkyWest Airlines Inc.  
336773  SkyWest Airlines Inc.  
336774  SkyWest Airlines Inc.  
336775  SkyWest Airlines Inc.  

[336776 rows x 20 columns]

In [ ]:
nycflights13.isnull().any()

In [ ]:
nycflights13.notnull()

In [ ]:
new_dataframe = nycflights13.dropna(axis = 0, how ='any')

In [ ]:
nycflights13.isnull().any()

### 11- Merge these two DataFrames into a single DataFrame called census.

Please use the following dataset for merging purposes:

'https://tf-assets-prod.s3.amazonaws.com/tf-curric/data-science/state-populations.csv'

'https://tf-assets-prod.s3.amazonaws.com/tf-curric/data-science/census-divisions.csv'

In [ ]:
state_populations = pd.read_csv('https://tf-assets-prod.s3.amazonaws.com/tf-curric/data-science/state-populations.csv')
census_divisions = pd.read_csv('https://tf-assets-prod.s3.amazonaws.com/tf-curric/data-science/census-divisions.csv')

In [ ]:
state_populations

In [ ]:
census_divisions

In [ ]:
census = pd.merge(state_populations, census_divisions)
census

### 12- Re-shape census so that one column contains all population measures, and another the year attributes.

In [ ]:
census.shape

In [ ]:
df = census.reset_index()
year_1 = pd.melt(census, id_vars=["state", "region", "division"],  value_vars=["2010", "2011", "2012", "2013", "2014", "2015", "2016"])
year_1

In [ ]:
year_2 = year_1.rename(columns={"variable": "year", "value": "population measures"})

In [ ]:
year_2

In [ ]:
grouped_by_year = year_2.groupby('year')
grouped_by_year.first()

### 15- Create a crosstab between state and category showing amount pledged as the value.

Please refer to the following link for this practice:

https://tf-assets-prod.s3.amazonaws.com/tf-curric/data-science/ksprojects.csv

In [27]:
ksproject=pd.read_csv(r"C:\Users\AAygun\Documents\GitHub\Data Sets\Magnimind\ksprojects.csv")

In [28]:
ksproject.head()

id                                               name  \
0  1147015301  "All We Had" Gets Into Cannes -- $10 or More G...   
1  1100844465                                       daily digest   
2  1316334968                             Drive A Faster Car 2.0   
3  1304906577         Accidental to Edinburgh - PHASE 1: AIRFARE   
4  1099226462          Logical Guess Pictures' 2nd Horror Movie!   

       category main_category             deadline    goal  \
0   Documentary  Film & Video  2009-05-20 21:50:00   300.0   
1   Documentary  Film & Video  2009-06-01 02:20:00   700.0   
2    Nonfiction    Publishing  2009-06-04 06:00:00  1000.0   
3       Theater       Theater  2009-06-05 05:59:00  6000.0   
4  Film & Video  Film & Video  2009-06-06 00:45:00   500.0   

              launched       state  backers country  usd_pledged currency  \
0  2009-04-30 22:10:30      failed        4      US        40.00      USD   
1  2009-05-04 21:14:28  successful       14      US       700.00      USD   
2  2009-05-04 22:51:31  successful       32      US      1367.00      USD   
3  2009-04-30 22:22:43  successful       24      US      6575.00      USD   
4  2009-04-30 01:32:55  successful       22      US       501.66      USD   

   pledged  
0    40.00  
1   700.00  
2  1367.00  
3  6575.00  
4   501.66

In [29]:
pd.crosstab(ksproject.state, ksproject.category)

category    3D Printing  Academic  Accessories  Action  Animals  Animation  \
state                                                                        
canceled             17        21           48      14        2         41   
failed               59        86          238      79       31        245   
live                  3         3           16       1        0          8   
successful           43        43          121      16       11        134   
suspended             2         4            2       0        1          1   
undefined             1         0            1       0        0          0   

category    Anthologies  Apparel  Apps  Architecture  ...  Wearables  Weaving  \
state                                                 ...                       
canceled              4      127   166            10  ...         25        3   
failed               34      684   784            86  ...         71        8   
live                  6       37    28             4  ...          7        0   
successful           83      211    55            26  ...         57        3   
suspended             0        5     5             2  ...          2        1   
undefined             0        0     0             0  ...          0        0   

category    Web  Webcomics  Webseries  Woodworking  Workshops  World Music  \
state                                                                        
canceled    114          4        119           10          0           31   
failed      634         28        656          125         12          174   
live         23          3          5            4          1            2   
successful   73         55        313           38          8          168   
suspended     6          0          6            2          1            1   
undefined     0          0          0            0          0            0   

category    Young Adult  Zines  
state                           
canceled             16      6  
failed               78     26  
live                  1      1  
successful           23     17  
suspended             0      0  
undefined             0      0  

[6 rows x 158 columns]